invertedai/large/initialize.py

```diff
@@ -0,0 +1,296 @@
+    will not be called if it is not possible for agents to exist there (e.g. there are no 
+    drivable surfaces present) or if the agent density criterion is already satisfied. As well,
+    predefined agents may be passed via the regions and will be considered as conditional. 
+
+    Arguments
+    ----------
+    location:
+        Location name in IAI format.
+
+    regions:
+        List of regions that contains information about the location, shape, and number of agents within the region.
+    
+    traffic_light_state_history:
+        History of traffic light states - the list is over time, in chronological order, i.e.
+        the last element is the current state. If there are traffic lights in the map, 
+        not specifying traffic light state is equivalent to using iai generated light states.
+
+    get_birdview:
+        If True, a birdview image will be returned representing the current world. Note this will significantly
+        impact on the latency.
+
+    get_infractions:
+        If True, infraction metrics will be returned for each agent.
+    
+    random_seed:
+        Controls the stochastic aspects of initialization for reproducibility.
+
+    api_model_version:
+        Optionally specify the version of the model. If None is passed which is by default, the best model will be used.
+
+    display_progress_bar:
+        If True, a bar is displayed showing the progress of all relevant processes.
+    
+    See Also
+    --------
+    :func:`initialize`
+    """
+
+    agent_states_sampled = []
+    agent_attributes_sampled = []
+    agent_rs_sampled = []
+
+    def inside_fov(center: Point, agent_scope_fov: float, point: Point) -> bool:
+        return ((center.x - (agent_scope_fov / 2) < point.x < center.x + (agent_scope_fov / 2)) and
+                (center.y - (agent_scope_fov / 2) < point.y < center.y + (agent_scope_fov / 2)))
+    
+    if display_progress_bar:
+        iterable_regions = tenumerate(
+            regions, 
+            total=len(regions),
+            desc=f"Initializing regions"
+        )
+    else:
+        iterable_regions = regions
+
+    for i, region in iterable_regions:
+        region_center = region.center
+        region_size = region.size
+
+        existing_agent_states, existing_agent_attributes, _ = _get_all_existing_agents_from_regions(regions)
+
+        conditional_agents = list(filter(
+            lambda x: inside_fov(center=region_center, agent_scope_fov=region_size+AGENT_SCOPE_FOV_BUFFER, point=x[0].center), 
+            zip(existing_agent_states,existing_agent_attributes)
+        ))
+
+        conditional_agent_states = [x[0] for x in conditional_agents]
+        conditional_agent_attributes = [x[1] for x in conditional_agents]
+
+        region_predefined_agent_states = [] if region.agent_states is None else region.agent_states
+        all_agent_states = conditional_agent_states + region_predefined_agent_states
```

Comment:
Aren't those already captured in `conditional_agent_states`? It looks like you're including them twice.

Reply:
The difference here is that `conditional_agent_states` contains states from regions that have already been initialized and predefined agents from other regions whereas `region_predefined_agent_states` contains predefined agents for the current region of interest. Any of these agents could be relevant as "conditional" for calling `initialize` on the current region.

Reply:
I've renamed `conditional_agent_states` to `out_of_region_conditional_agent_states` and used this naming scheme wherever else relevant.